docs/subnets/hello-world-precompile-tutorial.md

```diff
@@ -0,0 +1,946 @@
+	common.BytesToAddress([]byte{6}): &bn256AddIstanbul{},
+	common.BytesToAddress([]byte{7}): &bn256ScalarMulIstanbul{},
+	common.BytesToAddress([]byte{8}): &bn256PairingIstanbul{},
+	common.BytesToAddress([]byte{9}): &blake2F{},
+}
+```
+
+These precompile addresses start from `0x0000000000000000000000000000000000000001` and increment by 1.
+
+A [precompile](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/contracts.go#L53-L56) follows this interface:
+
+```go
+// PrecompiledContract is the basic interface for native Go contracts. The implementation
+// requires a deterministic gas count based on the input size of the Run method of the
+// contract.
+type PrecompiledContract interface {
+	RequiredGas(input []byte) uint64  // RequiredPrice calculates the contract gas use
+	Run(input []byte) ([]byte, error) // Run runs the precompiled contract
+}
+```
+
+Here is an example of the [sha256 precompile](https://github.com/ava-labs/subnet-evm/blob/master/core/vm/contracts.go#L238-L252) function.
+
+```go
+type sha256hash struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *sha256hash) RequiredGas(input []byte) uint64 {
+	return uint64(len(input)+31)/32*params.Sha256PerWordGas + params.Sha256BaseGas
+}
+
+func (c *sha256hash) Run(input []byte) ([]byte, error) {
+	h := sha256.Sum256(input)
+	return h[:], nil
+}
+```
+
+The CALL opcode (CALL, STATICCALL, DELEGATECALL, and CALLCODE) allows us to invoke this precompile.
+
+The function signature of CALL in the EVM is as follows:
+
+```go
+ Call(caller ContractRef,
+ addr common.Address,
+ input []byte,
+ gas uint64,
+ value *big.Int)
+ (ret []byte, leftOverGas uint64, err error)
+```
+
+Smart contracts in Solidity are compiled and converted into bytecode when they are first deployed. They are then stored on the blockchain and an address (usually known as the contract address) is assigned to it. When a user calls a function from a smart contract, it goes through the `CALL` function in the EVM. It takes in the caller address, the contract address, the input (function’s signature (truncated to the first leading four bytes) followed by the packed arguments data), gas, and value (native token). The function selector from the input lets the EVM know where to start from in the bytecode of the smart contract. It then executes a series of instructions (EVM opcodes) and returns the result.
+
+When a precompile function is called, it still goes through the `CALL` function in the EVM. However, it works a little differently. The EVM checks if the address is a precompile address from the mapping list and if so redirects to the precompile function.
+
+```go
+  if p := precompiles[addr]; p != nil {
+    return RunPrecompiledContract(p, input, contract)
+  }
+```
```

Comment:
Can we remove this example since this code is not actually present in the EVM? I think it would be better to keep this to a high-level explanation

```suggestion
```go
When a CALL  opcode is used, the EVM first checks if the address is a precompile address. If so, it executes the requested precompile. If not, it attempts to load a contract stored at that address.
```

Reply:
We should simplify this whole section about how a precompile is executed, we can keep it higher level than it is now

Reply:
@anusha-ctrl I cannot find what @aaronbuchwald suggested has been put into the doc. He suggested to add 

> When a CALL  opcode is used, the EVM first checks if the address is a precompile address. If so, it executes the requested precompile. If not, it attempts to load a contract stored at that address.

Reply:
Hi @yulin-dong, @aaronbuchwald and i worked on this section on a google meets together. thanks!